src/content/learn/typescript.md

```diff
@@ -0,0 +1,474 @@
+---
+title: Using TypeScript
+re: https://github.com/reactjs/react.dev/issues/5960
+---
+
+<Intro>
+
+TypeScript is a popular way to add type definitions to JavaScript codebases. Out of the box, TypeScript [supports JSX](/learn/writing-markup-with-jsx) and you can get full React Web support by adding [`@types/react`](https://www.npmjs.com/package/@types/react) and [`@types/react-dom`](https://www.npmjs.com/package/@types/react-dom) to your project.
+
+</Intro>
+
+<YouWillLearn>
+
+* [TypeScript with React Components](/learn/typescript#typescript-with-react-components)
+* [Typing common hooks](/learn/typescript#typing-hooks)
+* [Common types from `@types/react`](/learn/typescript/#useful-types)
+* [Further learning locations](/learn/typescript/#further-learning)
+
+</YouWillLearn>
+
+
+All of the production-grade frameworks mentioned in [Start a New React Project](/learn/start-a-new-react-project) offer support for using TypeScript with React, we recommend consulting their documentation for more information on setup. This guide assumes you have your project configured to support writing `*.tsx` TypeScript React files, and have finished the Quick Start guide.
+
+## TypeScript with React Components {/*typescript-with-react-components*/}
+
+Writing TypeScript with React is very similar to writing JavaScript with React. The key difference when working with a component is that you can provide types for your component's props. These types can be used for correctness checking and providing inline documentation in editors.
+
+Taking the [`MyButton` functional component](/learn#components) from the [Quick Start](/learn) guide, we can add a type describing the `title` for the button:
+
+<Sandpack>
+
+```tsx App.tsx active
+function MyButton({ title }: { title: string }) {
+  return (
+    <button>{title}</button>
+  );
+}
+
+export default function MyApp() {
+  return (
+    <div>
+      <h1>Welcome to my app</h1>
+      <MyButton title="I'm a button" />
+    </div>
+  );
+}
+```
+
+```js App.js hidden
+import AppTSX from "./App.tsx";
+export default App = AppTSX;
+```
+</Sandpack>
+
+ <Note>
+
+These sandboxes can handle TypeScript code, but they do not run the type-checker. This means you can amend the TypeScript sandboxes to learn, but you won't get any type errors or warnings. To get type-checking, you can use the [TypeScript Playground](https://www.typescriptlang.org/play) or use a more fully-featured online sandbox.
```

Comment:
What do we need in order to support this? It would be nice to have an example here where the type is wrong to see the benefit.

Reply:
I did look into it, and it's possible, https://github.com/codesandbox/sandpack/discussions/237#discussioncomment-3870506  preview: here https://codesandbox.io/s/github/danilowoz/sandpack-tsserver

It is a bit of a trade-off though, TypeScript is a ~6mb download for the client (.d.ts files and the compiler) and it's something that will occasionally want version bumping to keep up-to-date with the typescript runtime (though it could try re-use the version used by this repo) - all that for a few code samples is generally not worth it _IMO_.

For the TS website, I hardcoded all this information at build time by making https://shikijs.github.io/twoslash/ - but those code samples are static which I think is a bit restrictive WRT how the rest of the site works?